src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -0,0 +1,1355 @@
+                MDNS_EVENT_ENABLE_IP4 | MDNS_EVENT_ENABLE_IP6 |
+                MDNS_EVENT_ANNOUNCE_IP4 | MDNS_EVENT_ANNOUNCE_IP6);
+            ESP_ERROR_CHECK(mdns_netif_action(staIface_, action));
+            LOG(VERBOSE, "wifi: STA registered on mDNS.");
+        }
+    }
+#endif
+}
+
+//
+// EspIdfWiFi::wifi_event_handler()
+//
+void EspIdfWiFi::wifi_event_handler(esp_event_base_t base, int32_t id, void *data)
+{
+    if (base != WIFI_EVENT)
+    {
+        LOG(WARNING, "wifi: wifi_event_handler(), not a WIFI_EVENT.");
+        return;
+    }
+
+    switch (id)
+    {
+        default:
+            LOG(INFO, "wifi: wifi_event_handler() unknown id: %i", (int)id);
+            break;
+        case WIFI_EVENT_WIFI_READY:
+            started_ = true;
+            LOG(INFO, "wifi: WiFi started.");
+            break;
+        case WIFI_EVENT_STA_START:
+        {
+            std::string ssid;
+            std::string pass;
+            uint8_t authmode;
+            uint8_t channel;
+            last_sta_get(&ssid, &pass, &authmode, &channel);
+            if (ssid.empty())
+            {
+                // No valid "fast" connect parameters.
+                esp_wifi_scan_start(&SCAN_CONFIG, false);
+                LOG(INFO, "wifi: STA start, scanning...");
+            }
+            else
+            {
+                // Try the fast connect parameters first.
+                sta_connect(ssid, pass, authmode, channel);
+                LOG(INFO, "wifi: STA start, ssid: %s, pass: %s, connecting...",
+                    ssid.c_str(), pass.c_str());
+            }
+            break;
+        }
+        case WIFI_EVENT_STA_CONNECTED:
+        {
+            HASSERT(!connected_);
+            connected_ = true;
+            securityFailure_ = false;
+            wifi_event_sta_connected_t *sdata =
+                (wifi_event_sta_connected_t*)data;
+            LOG(INFO, "wifi: STA connected, channel: %u, authmode: %u",
+                sdata->channel, (unsigned)sdata->authmode);
+            last_sta_update(std::string((char*)sdata->ssid, sdata->ssid_len),
+                staConnectPass_, sdata->authmode, sdata->channel);
+            break;
+        }
+        case WIFI_EVENT_STA_DISCONNECTED:
+        {
+            HASSERT(connected_);
+            connected_ = false;
+            wifi_event_sta_disconnected_t *evdata =
+                (wifi_event_sta_disconnected_t *)data;
+            LOG(INFO, "wifi: STA disconnected ssid=%s reason %d rssi=%d",
+                evdata->ssid, evdata->reason, evdata->rssi);
+
+            if (evdata->reason == WIFI_REASON_AUTH_FAIL)
+            {
+                securityFailure_ = true;
+            }
+
+            bool try_fast_reconnect = false;
+            {
+                // We have to take a mutex because for some reason the
+                // IP_EVENT_STA_LOST_IP event does not get delivered when there
+                // is an unexpected WiFi disconnect. I guess it is implied?
```

Comment:
it's possible there is a timing / sequencing issue here, it may show up later as lost when disconnected.